src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs

```diff
@@ -0,0 +1,298 @@
+                            }
+                            else if (source[sourceIndex] == Utf8Constants.Minus)
+                            {
+                                offsetChar = Utf8Constants.Minus;
+                                state = Utf8Constants.DateJParserState.OffsetHours;
+                            }
+                            else
+							{
+								continueParsing = false;
+								break;
+							}
+
+                            sourceIndex++;
+                        }
+						break;
+					case Utf8Constants.DateJParserState.Fraction:
+						while (sourceIndex < sourceLength && ParserHelpers.IsDigit(source[sourceIndex]))
+						{
+                            if (!((fraction * 10) + (int)(source[sourceIndex] - 48u) <= Utf8Constants.MaxDateTimeFraction))
+                            {
+                                sourceIndex++;
+                                break;
+                            }
+
+                            fraction = (fraction * 10) + (int)(source[sourceIndex++] - 48u);
+						}
+
+						if (fraction != 0)
+						{
+							while (fraction * 10 <= Utf8Constants.MaxDateTimeFraction)
+								fraction *= 10;
+			            }
+
+						if (sourceIndex == sourceLength)
+							break;
+						
+						if (source[sourceIndex] == Utf8Constants.UtcOffsetChar)
+						{
+							offsetChar = Utf8Constants.UtcOffsetChar;
+							continueParsing = false;
+							sourceIndex++;
+						}
+						else if (source[sourceIndex] == Utf8Constants.Plus)
+                        {
+                            offsetChar = Utf8Constants.Plus;
+                            state = Utf8Constants.DateJParserState.OffsetHours;
+							sourceIndex++;
+                        }
+                        else if (source[sourceIndex] == Utf8Constants.Minus)
+                        {
+                            offsetChar = Utf8Constants.Minus;
+                            state = Utf8Constants.DateJParserState.OffsetHours;
+							sourceIndex++;
+						}
+                        else
+                            continueParsing = false;
+                        break;
+					case Utf8Constants.DateJParserState.OffsetHours:
+						state = ParserHelpers.TryGetNextTwoDigits(source, sourceIndex, out sourceIndex, out offsetHours) && (sourceLength - sourceIndex > 1) && (source[sourceIndex++] == Utf8Constants.Colon) ?
+                            Utf8Constants.DateJParserState.OffsetMinutes:
+                            Utf8Constants.DateJParserState.Invalid;
+						break;
+					case Utf8Constants.DateJParserState.OffsetMinutes:
+						if (!ParserHelpers.TryGetNextTwoDigits(source, sourceIndex, out sourceIndex, out offsetMinutes))
+                        {
+                            state = Utf8Constants.DateJParserState.Invalid;
+                            break;
+                        }
+
+                        continueParsing = false;
+						break;
+					case Utf8Constants.DateJParserState.Invalid:
+					default:
```

Comment:
What state would this catch? We may want to `Debug.Fail` for `default` since it likely indicates a bug in the code, no?